### ISSUE TITLE
ActionView is raising errors in the test environment after upgrading to Rails 6 (ActionView::Template::Error: undefined method `_app_views_xxxxx)

### ISSUE THREAD

**Issue by yjukaku** [rails/rails#40613]
### Expected behavior

ActionView template compilation is reliably executed in test and dev environments.

### Actual behavior

Tests (and sometimes the dev server) fail intermittently with exceptions like

```
Exception: ActionView::Template::Error: undefined method `_app_views_layouts__dmca_html_slim___3627166963330339270_108200' for #<#<Class:0x00007fac1acfaa58>:0x00007fac1ecec170>
Did you mean?  _app_views_layouts__header_nav_html_slim__4252095906685398667_109480
```

In our existing Rails app that we recently upgraded from 5.2.4 to 6.0.3.4, we're consistently facing build failures with the described exception. Rerunning a few times will sometimes make the issue go away.

### System configuration
**Rails version**:
Rails 6.0.3.4

**Ruby version**:
ruby 2.7.2p137

---

Unfortunately, I can't figure out steps to reproduce this on a new app :(, but I have done some diving into the ActionView code to try to figure out what's going on. 

I'm running a file with controller tests, and when I run the tests with a particular seed, I'm able to consistently reproduce the error.  Using pry-debugger and `puts`, I've worked out that ActionView template compilation seems to be the issue. For whatever reason, the `ActionView::Template` for the file has the `@compiled` flag set to `true`, but the `method_name` generated by the template has not been defined on the `view`.

When I step through the code here, https://github.com/rails/rails/blob/master/actionview/lib/action_view/template.rb#L152-L154 , I get to a state that I don't understand:

```ruby
    def render(view, locals, buffer = ActionView::OutputBuffer.new, add_to_stack: true, &block)
      instrument_render_template do
        compile!(view)
       # At this point, the template should have been compiled, and @compiled is set to true. But `view.respond_to?(method_name)` returns `false`. ??????
        view._run(method_name, self, locals, buffer, add_to_stack: add_to_stack, &block)
      end
    rescue => e
      handle_render_error(view, e)
    end
```
I've found two things that seem to stop the error.

First is adding a line before the `compile!` call above:

```ruby
    def render(view, locals, buffer = ActionView::OutputBuffer.new, add_to_stack: true, &block)
      instrument_render_template do
        @compiled = false
        compile!(view)
        view._run(method_name, self, locals, buffer, add_to_stack: add_to_stack, &block)
```

For obvious reasons, this forces ActionView to (re)compile the template, and this prevents the error from being raised.

Second, I've found that slowing down our first call to `render` seems to stop the issue from happening. In the controller test I mentioned above, I'm testing `FoosController#show`. `FoosController#show` explicitly calls `render "show"` (for unrelated reasons). If I insert a `sleep(0.1)` call right before the `render "show"` line, the issue disappears. If I then change that `sleep` call to `sleep(0.0)`, the error returns. I don't know what that means, but it seems like a race condition.

@jhawthorn , I know you were working on the template compilation recently. Do you have any ideas what could be happening here?

**Comment by jhawthorn** [comment]
Thanks for the detailed report. I appreciate the difficulty in making a simple reproduction.

This seems a lot like https://github.com/rails/rails/pull/37928, which should have been fixed in 6.0.2+, but this could be something similar.

It's surprising to hear it fails in tests, where I would generally expect fewer concurrency issues.

**Comment by yjukaku** [comment]
Thanks for that link. I didn't find anything earlier when searching for similar errors.

I was able to confirm locally that it seems to be an issue with the cache expiry. Here's what I did:

1. Added instrumentation to our tests to `puts 'rendered template'` whenever `"!render_template.action_view"` was called.
2. `bundle open actionview` to edit `cache_expiry.rb` with the following three changes:
```ruby

    def clear_cache_if_necessary
      @mutex.synchronize do
        puts "Entered mutex" # Change 1
        watched_dirs = dirs_to_watch
        next if watched_dirs.empty? # Change 2 (changed `return` to `next` so that the "Left mutex" line would always be called)

        if watched_dirs != @watched_dirs
          @watched_dirs = watched_dirs
          @watcher = @watcher_class.new([], watched_dirs) do
            clear_cache
          end
          @watcher.execute
        else
          @watcher.execute_if_updated
        end
      end
      puts "Left mutex" # Change 3
    end
```
3. Ran the tests that were failing earlier.

Here's what I found: As expected, after all of the views were rendered and a bunch of `"rendered template"` were output to the console, the text "Entered mutex" was output. RSpec displayed that the first test passed successfully.

However, the next part is strange. Before the "Left mutex" was displayed, there were a lot more `"rendered template"` lines output.

So it seems that the mutex for entering the code `clear_cache_if_necessary` is not preventing *new* templates from being compiled while it gets cleared. That makes sense - the mutex is not shared across the insertion into the cache and clearing of the cache. 

I'm not sure why our team seems to be the only one on the interwebs facing this issue, though. We don't have a particular huge app. Maybe something else in our codebase is making the `clear_cache_if_necessary` call slower than most apps.

Does my analysis make sense to you?

**Comment by jhawthorn** [comment]
Thanks! There's a few interesting things here, which might help explain why there aren't more folks experiencing this and might help us find a reproduction outside of your app.

I'm guessing you have `config.cache_classes = false` but maybe not `config.action_view.cache_template_loading = true` in `config/environments/test.rb`? I feel like in most cases we shouldn't want/need to be checking view expiry from tests. (IMO it's a bug that we get errors regardless of configuration, but this might help narrow down the needed conditions for this failure)

Is it possible `#append_view_path` is being used? That seems to be a common source of problems. (similarly, IMO a bug if it causes problems, but would narrow down the issue)

The thing I'd most like to know is where these renders are coming from (standard requests? a mailer? system tests?). It seems a bit odd that we're getting a few `rendered template` from the first test without first going through the cache expiry, but then we do start running the cache expiry before that first test ends, but it continues running when when the test finishes, which is surprising. There's clearly multiple threads so I'd like to know what they are.

**Comment by yjukaku** [comment]
> I'm guessing you have config.cache_classes = false but maybe not config.action_view.cache_template_loading = true in config/environments/test.rb? I feel like in most cases we shouldn't want/need to be checking view expiry from tests.

We do have `config.cache_classes = false` in our `test.rb`, but we don't have anything for `config.action_view.cache_template_loading`. However, [according the Rails docs](https://guides.rubyonrails.org/configuring.html#configuring-action-view), that value should default to the same as  `config.cache_classes` (false), and it does when I check the value of `ActionView::Resolver.caching?` during a test run.

On the other hand, setting it to `true` does seem to make the error go away.

> Is it possible #append_view_path is being used? That seems to be a common source of problems. (similarly, IMO a bug if it causes problems, but would narrow down the issue)

I don't see use of `append_view_path` (or `prepend_view_path`) in our codebase.

> The thing I'd most like to know is where these renders are coming from (standard requests? a mailer? system tests?)

These are coming from
```
> FoosController
> #show
> call to render "show"
> application.html.slim
```

Our `application.html.slim` renders a lot of partials, and at some point while rendering those partials, the error is raised.

> It seems a bit odd that we're getting a few rendered template from the first test without first going through the cache expiry, but then we do start running the cache expiry before that first test ends, but it continues running when when the test finishes, which is surprising. There's clearly multiple threads so I'd like to know what they are.

Sorry, I don't follow. 🤔  AFAIU, the cache expiry should happen after the request finishes, right? So it's expected that we see templates rendered _and then_ the cache expiry.

I agree it's surprising that a new request can start while the cache is getting expired. That's really strange and it would indicate that rspec is running the next test in a new thread, right? That would be really unexpected. We don't have any special rspec config which could cause that to happen, AFAICT.

**Comment by yjukaku** [comment]
OK looks like the test runner is running each test in the same thread as expected 😅 . 

But it looks like the `clear_cache_if_necessary` call is happening in a different thread than the test runner.

```
FoosController
  POST #one_action
    test passed
    test passed
  POST #another_action
    test passed
  GET #show
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
clearing cache... in thread with object_id 105640
    name of some test
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
cache cleared.
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
clearing cache... in thread with object_id 107960
    name of another test
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
template rendered in thread with object_id 16820
```

**Comment by jhawthorn** [comment]
> Sorry, I don't follow. 🤔 AFAIU, the cache expiry should happen after the request finishes, right? So it's expected that we see templates rendered _and then_ the cache expiry.

Cache expiry should happen first, before any templates are rendered.

**Comment by yjukaku** [comment]
> Cache expiry should happen first, before any templates are rendered.

Ah, I misunderstood that. So the output above is very confusing indeed. It's as if the `to_run` hook for the executor is getting executed as an `to_complete` hook.

In any case, we changed our `config/environments/test.rb` to have `config.action_view.cache_template_loading = true`, so hopefully our problem is resolved. If you need any more help debugging, please let me know, but I don't know how much more time I can spend on this if the problem goes away for our app.

Thanks so much for your help so far! This was driving me crazy 😅

**Comment by yjukaku** [comment]
I was able to repro in development mode as well using a new Rails app:

https://github.com/yjukaku/RailsTemplateCompilationErrorRepro

Simply run the `bin/trigger_template_cache_error` script after starting the server at port 3001.

The issue in development can be caused by modifying a partial while requests are coming in. This is kind of an edge case in development but it happens to me personally because we have webpacker set up to automatically refresh on file changes, and I usually have multiple Chrome tabs open pointing at localhost. When a refresh is triggered on all 3 tabs, I often am continuing to make changes to template code 😓 , which seems to be what triggers this error.

**Comment by caifara** [comment]
We also ran against this problem. We do use `append_view_path` and have apparently seen the error in production (I've found one occasion in sentry, but the app isn't used much (yet, I hope)). Running Rails 6.0.3.4.

Haven't seen it in test environment however.

**Comment by yjukaku** [comment]
@caifara , what do you have set for `config.cache_classes`and `config.action_view.cache_template_loading` in your `config/environments/production.rb`?

**Comment by caifara** [comment]
I may be seeing two problems, but at least one of them is not this, but a problem in view_component: https://github.com/github/view_component/pull/538

**Comment by caifara** [comment]
@yjukaku I don't think the problem is related to production config (set to true and default (not in config)), but did find a problem in view_component. With that fix in place, the errors seem to be gone. Thanks for the help!

**Comment by woodhull** [comment]
I think we may also be experiencing this. We're prepending view paths per request from a rails engine to allow templates to be overridden for specific customers which might be related.

**Comment by woodhull** [comment]
More info on our particular issue:

The crash has been happening for us in development mode. We think it's something about the fact that we're using prepend_view_path on a per request basis, since commenting this out resolves the issue. The crash is easy to replicate. App works until there is any template filesystem change, and then crashes when it can not find the template the application thinks should be compiled from the `prepend_view_path`

**Comment by koppen** [comment]
Not sure if more +1's is helpful here, but we're experiencing this as well:

- rails (6.0.3.4)
- `Rails.env #=> "development"`
- `config.cache_classes = false`
- `config.action_view.cache_template_loading` isn't set (`Rails.application.config.action_view.cache_template_loading #=> nil`)
- We have no `append_view_path` config.
- We have no `prepend_view_path` config.
- We are running `view_component (2.28.0)` - not sure if this is relevant, but it was mentioned above. We have seen the issue on another app not using view_component, though.

We're only seeing this occasionally, but at least anecdotally @yjukaku's [explanation above](https://github.com/rails/rails/issues/40613#issuecomment-729035449) matches our experiences as well; modifying a template while a request is being processed seem to trigger the error.

**Comment by mfcorgi** [comment]
This is also happening to me when running the tests. on 6.0.3.6 -The exception is something like this,

# --- Caused by: ---
     # NoMethodError:
     #   undefined method `_app_views_XXXXXXXXXXX__4472450733415361230_47313648806920' for #<ActionView::Base:0x0000561022862e40>
     #   Did you mean?  _app_views__app_views_XXXXXXXXXXX__4472450733415361230_47313645039900
     #                  _app_views__app_views_XXXXXXXXXXX__4472450733415361230_47313624130720
     #                  _app_views__app_views_XXXXXXXXXXX__4472450733415361230_47313636373300
     #   /usr/local/bundle/gems/actionview-6.0.3.6/lib/action_view/base.rb:274:in `_run'

**Comment by davidwessman** [comment]
I have been struggling with this error for a while now working on an upgrade from Rails 5.2.6 -> 6.0.4.

So far I have only seen it in tests, and in test.rb I have:
- `config.cache_classes = true`
- `config.action_view.cache_template_loading = true`

This is the only mentions I have found about similar issues.
Does anyone have any hints about how to solve this? Should I open a new issue?

I can see that the fix by @jhawthorn can only be found on the main branch: #41198

**Comment by davidwessman** [comment]
> I have been struggling with this error for a while now working on an upgrade from Rails 5.2.6 -> 6.0.4.
> 
> So far I have only seen it in tests, and in test.rb I have:
> 
> * `config.cache_classes = true`
> * `config.action_view.cache_template_loading = true`
> 
> This is the only mentions I have found about similar issues.
> Does anyone have any hints about how to solve this? Should I open a new issue?
> 
> I can see that the fix by @jhawthorn can only be found on the main branch: [commit](https://github.com/rails/rails/pull/41998/commits/9a4c1e205ee5c4facef3c391e2a63bc3cff7d7f8)

I think I found the culprit:

The same partial was rendered via a normal controller and via a custom ActionView::Base call without a controller context (which is deprecated in Rails 6), the first time the partial is called it will cache the result as a method that is no longer defined in the second context.

**Comment by GCorbel** [comment]
I have a similar issue but in development only and have no issue on test. In develpoment.rb there is, `config.cache_classes = false` and nothing for `config.action_view.cache_template_loading`. I have no deprecation warning about the lookup context. The method `prepend_view_path` is used. The error happens when a view file is changed.

I tried to set `config.action_view.cache_template_loading = true` and it fixes the issue, but changes in view files aren't compiled. The server have to be restarted after each change.

The dirty hack I found at the moment is to simply always compile the template :

```
module AlwaysCompile
  def compile!(_view)
    @compiled = false if Rails.env.development?
    super
  end
end

ActionView::Template.prepend(AlwaysCompile)
```

**Comment by zhengpd** [comment]
Similar issue when using `ActionView::Base` after upgrading to rails 6.1:

```ruby
view = ActionView::Base.with_empty_template_cache.with_view_paths(paths)
view.render(template: 'label_styles/show')
# => undefined method _app_views_xxxx
```

After switched to a subclass with `compiled_method_container`, the error is gone:

```ruby
class LabelStyleView < ActionView::Base
  def compiled_method_container
    self.class
  end
end
view = LabelStyleView.with_view_paths(paths)
view.render(template: 'label_styles/show')
# => works fine
```

I guess @davidwessman is right about the culprit: 

> the first time the partial is called it will cache the result as a method that is no longer defined in the second context

**Comment by davidwessman** [comment]
> So does anyone here have a clean and clear fix for this?

Do not call ActionView::Base directly is my only tip.

**Comment by headshota** [comment]
This might be helpful for someone, we had similar issue, but this was related to spec type mixing in files, while having `config.infer_spec_type_from_file_location!`

We had several `type: :request` specs in controller subdirectory, which was causing similar errors. Moving these specs to dedicated `spec/requests` subdirectory solved the issue.

**Comment by simi** [comment]
Just faced the same problem with Rails 7.0.4. Adding `config.action_view.cache_template_loading = true` to `test.rb` fixed this. @jhawthorn should https://github.com/rails/rails/pull/41998 fix this to work also without that config? If so, problem still persists. I can try to debug if needed.

**Comment by amolbudkhe** [comment]
We are also facing this issue, but not for the test cases. We are facing this issue with report generation. We are using the sidekiq gem for generating the Excel files. After upgrading to rails 6.1.7 we are facing this issue. The irony is that when we deploy the sidekiq, then we are able to generate a report only one time, after getting generated we start getting such errors. After restarting the sidekiq again the report will generate one time, after that same error will occur.


Can anyone help us with this?

Should I create a new issue for this? 

Error snippet.
```
2023-02-02T09:21:20.955Z 8 TID-1puw GenerateReportTask JID-dce45b05122f7c01039d4c9e INFO: fail: 3.678 sec
2023-02-02T09:21:20.955Z 8 TID-1puw WARN: {"context":"Job raised exception","job":{"class":"GenerateReportTask","args":[{"report_type":"report_cr_profiles","entities":{"cr_profile_id":"238"}}],"retry":false,"queue":"reports","backtrace":true,"jid":"dce45b05122f7c01039d4c9e","created_at":1675329677.2719676,"enqueued_at":1675329677.2720017},"jobstr":"{\"class\":\"GenerateReportTask\",\"args\":[{\"report_type\":\"report_cr_profiles\",\"entities\":{\"cr_profile_id\":\"238\"}}],\"retry\":false,\"queue\":\"reports\",\"backtrace\":true,\"jid\":\"dce45b05122f7c01039d4c9e\",\"created_at\":1675329677.2719676,\"enqueued_at\":1675329677.2720017}"}
2023-02-02T09:21:20.955Z 8 TID-1puw WARN: ActionView::Template::Error: undefined method `_app_views_reports_report_cr_profiles_index_xlsx_axlsx__1412410369129045891_139960' for #<ActionView::Base:0x00000000084b98>
2023-02-02T09:21:20.955Z 8 TID-1puw WARN: /var/lib/gems/2.7.0/gems/actionview-6.1.7.1/lib/action_view/base.rb:247:in `public_send'
/var/lib/gems/2.7.0/gems/actionview-6.1.7.1/lib/action_view/base.rb:247:in `_run'
/var/lib/gems/2.7.0/gems/actionview-6.1.7.1/lib/action_view/template.rb:154:in `block in render'
```

**Comment by davidwessman** [comment]
@amolbudkhe Can you include a snippet of how you render the view?
We switched all rendering in workers to call `ExcelController.render()` and we have no more issues.

**Comment by pt** [comment]
@davidwessman in case it's helpful, I replaced some janky code rendering templates in a delayed job with a call to the controller to render the template and it fixed this problem for me. Thanks for the comment.

**Comment by iainbeeston** [comment]
I'm finding this too in the test environment but only on my development machine (CI works fine). I have:

* `rails (7.0.8.4)`
* `rspec-rails (6.1.3)`
* `rspec-core (3.13.0)`
* `view_component (3.12.1)`
* `config.enable_reloading = false`
* `config.action_view.cache_template_loading` isn't set
* We're not changing the view path (append or prepend)

Setting `config.action_view.cache_template_loading = true` fixes the issue
